### PR TITLE
[spirv] Support translation of Buffer & RWBuffer.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp [hunk 1]
```diff
@@ -622,13 +622,37 @@ uint32_t ModuleBuilder::getFunctionType(uint32_t returnType,
...
-                                    /*sampled*/ 1, spv::ImageFormat::Unknown);
+                                     uint32_t depth, bool isArray, uint32_t ms,
+                                     uint32_t sampled,
+                                     spv::ImageFormat format) {
+  const Type *type = Type::getImage(theContext, sampledType, dim, depth,
+                                    isArray, ms, sampled, format);
   const uint32_t typeId = theContext.getResultIdForType(type);
   theModule.addType(type, typeId);
 
+  if (format == spv::ImageFormat::Rg32f || format == spv::ImageFormat::Rg16f ||
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: A `switch` statement will be better?

File: tools/clang/lib/SPIRV/TypeTranslator.cpp [hunk 2]
```diff
@@ -450,9 +450,47 @@ uint32_t TypeTranslator::translateResourceType(QualType type) {
...
+  // Buffer and RWBuffer types
+  if (name == "Buffer" || name == "RWBuffer") {
+    theBuilder.requireCapability(spv::Capability::SampledBuffer);
+    const auto sampledType = hlsl::GetHLSLResourceResultType(type);
+    const auto format = translateSampledTypeToImageFormat(sampledType);
+    return theBuilder.getImageType(
+        translateType(getElementType(sampledType)), spv::Dim::Buffer,
+        /*depth*/ 0, /*isArray*/ 0, /*ms*/ 0,
+        /*sampled*/ name == "Buffer" ? 1 : 2,
+        translateSampledTypeToImageFormat(sampledType));
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.cpp]: this is `format` defined in the above?

File: tools/clang/lib/SPIRV/TypeTranslator.h [hunk 3]
```diff
@@ -144,6 +144,10 @@ class TypeTranslator {
   std::pair<uint32_t, uint32_t>
   getAlignmentAndSize(QualType type, uint32_t *stride, bool isRowMajor);
 
+  /// \bried For the given sampled type, returns the corresponding image format
+  /// that can be used to create an image object.
+  spv::ImageFormat translateSampledTypeToImageFormat(QualType type);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Can be `static` method

File: tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp [hunk 4]
```diff
@@ -450,5 +450,6 @@ TEST_F(FileTest, ComputeByteAddressBufferLoad) {
 TEST_F(FileTest, ComputeByteAddressBufferStore) {
   runFileTest("compute.byte-address-buffer.store.hlsl");
 }
+TEST_F(FileTest, ComputeBuffer) { runFileTest("compute.buffer.hlsl"); }
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/CodeGenSPIRVTest.cpp]: This belongs to the `type` category. Preferably should be named as `type.buffer.hlsl`

File: tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+// CHECK: %float4buf = OpVariable %_ptr_UniformConstant_type_buffer_image_13 UniformConstant
+// CHECK: %int3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_14 UniformConstant
+// CHECK: %int4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_14 UniformConstant
+// CHECK: %uint3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_15 UniformConstant
+// CHECK: %uint4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_15 UniformConstant
+// CHECK: %float3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_16 UniformConstant
+// CHECK: %float4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_16 UniformConstant
+
+[numthreads(1, 1, 1)]
+void main() {}
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl]: Oh, BTW, we should use graphics shaders for testing. Using images in compute shaders requires additional capabilities.

Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl]: What capabilities?
do you mean I should change this test to be a PS, or i should keep this and also add a PS test?

Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl]: Yes. For capabilities, I cannot recall right now. Maybe I just somehow got the wrong impression. But images are typically used in graphics shaders.

**Comment by antiagainst** [comment]
And we also need to make sure we have correct binding support for them: adding variables for them in `vk.binding.{explicit|implicit|register}.hlsl`. (Not necessarily for this pull request)

**Comment by antiagainst** [comment]
If we cannot annotate them with `[[vk::binding()]``, just leave a TODO in the test later we can revisit.

File: tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl (hunk 5 above)
Comment (ehsannas) [inline tools/clang/test/CodeGenSPIRV/compute.buffer.hlsl]: done

**Comment by ehsannas** [comment]
I'll try `[[vk::binding())]]` in another CL

File: tools/clang/lib/SPIRV/TypeTranslator.h (hunk 3 above)
Comment (antiagainst) [inline tools/clang/lib/SPIRV/TypeTranslator.h]: Oh, I see.

File: tools/clang/test/CodeGenSPIRV/type.buffer.hlsl [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+// CHECK: %float3buf = OpVariable %_ptr_UniformConstant_type_buffer_image_13 UniformConstant
+// CHECK: %float4buf = OpVariable %_ptr_UniformConstant_type_buffer_image_13 UniformConstant
+// CHECK: %int3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_14 UniformConstant
+// CHECK: %int4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_14 UniformConstant
+// CHECK: %uint3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_15 UniformConstant
+// CHECK: %uint4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_15 UniformConstant
+// CHECK: %float3rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_16 UniformConstant
+// CHECK: %float4rwbuf = OpVariable %_ptr_UniformConstant_type_buffer_image_16 UniformConstant
+
+[numthreads(1, 1, 1)]
```
Comment (antiagainst) [inline tools/clang/test/CodeGenSPIRV/type.buffer.hlsl]: Pixel shader does not support `numthreads`. :)

**Pull request merged**